mark.c

```diff
@@ -864,40 +895,45 @@ GC_INNER mse * GC_mark_from(mse *mark_stack_top, mse *mark_stack,
           if (cheri_gettag(limit) == 0) {
             limit -= ALIGNMENT;
           } else {
-            deferred = *(word *)limit;
+            deferred = *(void **)limit;
             FIXUP_POINTER(deferred);
             limit -= ALIGNMENT;
             if (deferred >= (word)least_ha && deferred < (word)greatest_ha)
               break;
           }
           if ((word)current_p > (word)limit) goto next_object;
         }
-#       endif   // defined(__CHERI_PURE_CAPABILITY__)
-#     endif
 
-      while ((word)current_p <= (word)limit) {
-        /* Empirically, unrolling this loop doesn't help a lot. */
-        /* Since PUSH_CONTENTS expands to a lot of code,        */
-        /* we don't.                                            */
-        current = *(word *)current_p;
-        FIXUP_POINTER(current);
-        PREFETCH(current_p + PREF_DIST*CACHE_LINE_SIZE);
-        if (current >= (word)least_ha && current < (word)greatest_ha) {
-          /* Prefetch the contents of the object we just pushed.  It's  */
-          /* likely we will need them soon.                             */
-          PREFETCH((ptr_t)current);
-#         ifdef ENABLE_TRACE
-            if (GC_trace_addr == current_p) {
-              GC_log_printf("GC #%u: considering(1) %p -> %p\n",
-                            (unsigned)GC_gc_no, (void *)current_p,
-                            (void *)current);
-            }
-#         endif /* ENABLE_TRACE */
-          PUSH_CONTENTS((ptr_t)current, mark_stack_top,
-                        mark_stack_limit, current_p);
+
+        while ((word)current_p <= (word)limit) {
+          /* Empirically, unrolling this loop doesn't help a lot. */
+          /* Since PUSH_CONTENTS expands to a lot of code,        */
+          /* we don't.                                            */
+          current = *(void **)current_p;
+          if (cheri_gettag(current) == 0) {
```

Comment:
For completeness this _could_ also exclude capabilities that don't have at least one of RWX (since they are required to address memory). There's a trade-off there: with today's workloads, pretty much all the capabilities address memory, so these checks are just overhead. It might be a tuning option to experiment with later, though.

Reply:
> For completeness this could also exclude capabilities that don't have at least one of RWX

I'm not sure we want to do this. For example, someone might hand out capabilities without RWX to an address knowing that they can convert them back into a readable/writable address later. Put another way: I agree that checking for one of RWX is overhead, and doing so probably makes the system slightly more fragile :)

Reply:
> knowing that they can convert them back into a readable/writable address later

Oh, I'm not so sure about that. The extra check here allows the GC to identify memory pointers with certainty, making it precise. A capability without R|W|X is not a pointer to memory, and a precise GC should not mark it. Without the R|W|X check, we'd probably have to call it "nearly precise".

In the example you gave: the only way to restore a permission to a capability would be to re-derive it from another, superior capability, and in that case the GC should be able to find and mark that one.

Reply:
> A capability without R|W|X is not a pointer to memory

Ah, that's interesting: is that part of the formal spec of CHERI capabilities (as such)? If so, I agree with you.

Reply:
I had simply inferred it from the observation that every interaction with memory (including instruction fetches) requires at least one of those permissions. However, [CHERI v8 (TR-951)](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-951.pdf) has some words that might help (under "2.3.9 Capability Protection for Non-Pointer Types"):

> If the mask authorizes load, store, and fetch instructions, then the capability has a pointer interpretation. Capabilities are not required to have those permissions set, however, allowing capabilities to be used for other purposes – for example, [...]

For Morello specifically, I think the "Capability architecture rules" from the [Morello architecture supplement](https://developer.arm.com/documentation/ddi0606/latest) also encode this property. For example, rules R<sub>GMYFJ</sub>, R<sub>HMXNK</sub> and R<sub>YPPQB</sub> together seem to cover this for loads and stores, and R<sub>ZZWCP</sub> for instruction fetchs. Other rules describe permission requirements for less obvious memory-addressing operations, such as cache maintenance.

Reply:
Right, this definitely backs your point up: "If the mask authorizes load, store, and fetch instructions, then the capability has a pointer interpretation".

@djichthys A bit up to you whether you want to include this check or not. I'm not entirely sure how expensive it might be (and we don't really, I think, have a good way of benchmarking at the moment), but it might be better to put it in now (while we remember) and take it out later if we find that it's too expensive.

Reply:
> @djichthys A bit up to you whether you want to include this check or not. 

Check inserted in [34b820e](https://github.com/capablevms/bdwgc/pull/13/commits/34b820e291036646a084ff140f33c662801fa200) .